microprofile-fault-tolerance/src/test/java/org/jboss/eap/qe/microprofile/fault/tolerance/v10/MicroProfileFaultTolerance10Test.java

```diff
@@ -0,0 +1,398 @@
+
+        // 20 = 10 allowed invocations + 10 queued invocations
+        // 21 invocations would already trigger fallback
+        testBulkhead(20, baseApplicationUrl + "async?operation=bulkhead", expectedResponses);
+    }
+
+    /**
+     * @tpTestDetails Deploy MP FT application with @BulkHead, @Asynchronous and @Fallback annotation on service method.
+     *                Bulkhead allows max 10 requests + 10 queued. In parallel call 40 requests.
+     * @tpPassCrit Verify 20 requests passed @BulkHead method and 20 went to @Fallback method
+     * @tpSince EAP 7.4.0.CD19
+     */
+    @Test
+    public void bulkheadFailureAsync() throws InterruptedException {
+        Map<String, Integer> expectedResponses = new HashMap<>();
+        expectedResponses.put("Hello from @Bulkhead method", 20);
+        expectedResponses.put("Fallback Hello", 20);
+
+        // 40 = 10 allowed invocations + 10 queued invocations + 20 not allowed invocations that lead to fallback
+        // 41 invocations would already trigger fallback rejection
+        testBulkhead(40, baseApplicationUrl + "async?operation=bulkhead&fail=true", expectedResponses);
+    }
+
+    private static void testBulkhead(int parallelRequests, String url, Map<String, Integer> expectedResponses)
+            throws InterruptedException {
+        Set<String> violations = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        Queue<String> seenResponses = new ConcurrentLinkedQueue<>();
+
+        ExecutorService executor = Executors.newFixedThreadPool(parallelRequests);
+        for (int i = 0; i < parallelRequests; i++) {
+            executor.submit(() -> {
+                try {
+                    String response = get(url).asString();
+                    seenResponses.add(response);
+                } catch (Exception e) {
+                    violations.add("Unexpected exception: " + e.getMessage());
+                }
+            });
+        }
+        executor.shutdown();
+        boolean finished = executor.awaitTermination(10, TimeUnit.SECONDS);
+        assertThat(finished, is(true));
+
+        for (String seenResponse : seenResponses) {
+            if (!expectedResponses.containsKey(seenResponse)) {
+                violations.add("Unexpected response: " + seenResponse);
+            }
+        }
+
+        for (Map.Entry<String, Integer> expectedResponse : expectedResponses.entrySet()) {
+            int count = 0;
+            for (String seenResponse : seenResponses) {
+                if (expectedResponse.getKey().equals(seenResponse)) {
+                    count++;
+                }
+            }
+            if (count != expectedResponse.getValue()) {
+                violations.add("Expected to see " + expectedResponse.getValue() + " occurence(s) but seen " + count
+                        + ": " + expectedResponse.getKey());
+            }
+        }
+
+        assertThat(violations, is(empty()));
```

Comment:
Is there any reason why not fail immediately when unexpected state is recognized?

Reply:
Yes, this is used for tests with `@Bulkhead` where there are multiple parallel calls. Failing with first failed call would not provide overall status how parallel calls finished and what actually happened.

Also as all calls are fired at once(in parallel) there would be no real benefit when failing with first.

Reply:
Thanks for explanation.